lib/teiserver/o_auth.ex

```diff
@@ -0,0 +1,419 @@
+  end
+
+  defp equal_uri?(uri1, uri2) do
+    equal_uri?(URI.parse(uri1), uri2)
+  end
+
+  # https://www.rfc-editor.org/rfc/rfc8252#section-7.3
+  # although `localhost` is not recommended, it is allowed, see:
+  # and https://www.rfc-editor.org/rfc/rfc8252#section-8.3
+  defp localhost?(%URI{} = uri), do: localhost?(uri.host)
+  defp localhost?("localhost"), do: true
+  defp localhost?("127.0.0.1"), do: true
+  defp localhost?("::1"), do: true
+  defp localhost?("0:0:0:0:0:0:0:1"), do: true
+  defp localhost?(_), do: false
+
+  @doc """
+  Create an authorization token for the given user and application.
+  The token scopes are the same as the application
+  """
+  @spec create_code(
+          User.t() | T.userid(),
+          %{
+            id: integer(),
+            scopes: Application.scopes(),
+            redirect_uri: String.t(),
+            challenge: String.t(),
+            challenge_method: String.t()
+          },
+          options()
+        ) ::
+          {:ok, Code.t()} | {:error, Ecto.Changeset.t()}
+  def create_code(user, attrs, opts \\ [])
+
+  def create_code(%User{} = user, attrs, opts) do
+    create_code(user.id, attrs, opts)
+  end
+
+  def create_code(user, attrs, opts) when is_map(user) do
+    create_code(user.id, attrs, opts)
+  end
+
+  def create_code(user_id, attrs, opts) do
+    now = Keyword.get(opts, :now, Timex.now())
+
+    # don't do any validation on the challenge yet, this is done when exchanging
+    # the code for a token
+    attrs = %{
+      value: Base.hex_encode32(:crypto.strong_rand_bytes(32)),
+      owner_id: user_id,
+      application_id: attrs.id,
+      scopes: attrs.scopes,
+      expires_at: Timex.add(now, Timex.Duration.from_minutes(5)),
+      redirect_uri: Map.get(attrs, :redirect_uri),
+      challenge: Map.get(attrs, :challenge),
+      challenge_method: Map.get(attrs, :challenge_method)
+    }
+
+    %Code{}
+    |> Code.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Given a code returns the corresponding db object, making sure
+  it is valid (exists, not expired, not revoked)
+  """
+  @spec get_valid_code(String.t(), options()) :: {:ok, Code.t()} | {:error, term()}
+  def get_valid_code(code, opts \\ [])
+
+  def get_valid_code(code, opts) do
+    case CodeQueries.get_code(code) do
+      nil ->
+        {:error, :no_code}
+
+      code ->
+        now = Keyword.get(opts, :now, Timex.now())
+        check_expiry(code, now)
+    end
+  end
+
+  @spec create_token(
+          User.t() | T.userid(),
+          %{id: integer(), scopes: Application.scopes()},
+          create_refresh: boolean() | options()
+        ) ::
+          {:ok, Token.t()} | {:error, Ecto.Changeset.t()}
+  def create_token(user_id, application, opts \\ [])
+
+  def create_token(%User{} = user, application, opts) do
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user, application, opts) when is_map(user) do
+    create_token(user.id, application, opts)
+  end
```

Comment:
It's used for testing mostly, because the [new_user/2](https://github.com/geekingfrog/teiserver/blob/master/test/support/teiserver_test_lib.ex#L46) function to create test users doesn't return an ecto schema.

And I believe in general, there aren't many case where teiserver directly handle `User` structs. This makes it a bit more robust, especially since we only care about the id.

As a side note, this is something that I'm not a big fan of: the way users are augmented with all sort of data, makes it quite hard to know what's persisted in the DB (and how) from what's transient.